Rest API приложение библиотеки на Spring Boot

ITWeiss opened the pull request:
- Реализована возможность регистрации пользователя, авторизация
- Реализованы уровни доступа с помощью Spring Security
- Реализованы репозитории, сервисы и контроллеры
- Заполнение бд реализовано через миграции с использованием библиотеки flyway
- Добавлены тесты
- Добавлена проверка стиля кода
- Добавлены java doc

ITWeiss committed:
Rest API приложение библиотеки на Spring Boot
- Реализована возможность регистрации пользователя, авторизация
- Реализованы уровни доступа с помощью Spring Security
- Реализованы репозитории, сервисы и контроллеры
- Заполнение бд реализовано через миграции с использованием библиотеки flyway
- Добавлены тесты
- Добавлена проверка стиля кода
- Добавлены java doc

RinatBeybutov commented on src/main/java/com/example/libraryProject/entity/Book.java:
```diff
@@ -0,0 +1,48 @@
+  private String updatedPerson;
+  private String removedPerson;
+
+  @ManyToOne(fetch = FetchType.LAZY)
```
почему тут выбрано lazy?

ITWeiss replied:
Уточняла у Ивана когда начинала этот проект, он сказал что нужно указать FetchType.LAZY
Нужно указать EAGER или вообще не указывать?

RinatBeybutov replied:
Краткая справка: Если у тебя поле - один объект, там всегда по умолчанию Eager. А если у тебя поле - список, то будет по умолчанию Lazy.

Выбор нужно типа зависит от ситуации:
Если не хочешь, чтобы грузились эти пользователи, то оставь так. Но в запросе, где этот пользователь понадобится в репозитории нужно будет join fetch использовать, иначе встретишь проблему n+1. 

От одного поля проблем не будет, тк у тебя по внешнему ключу быстро запись найдется. Если будет два десятка полей, то можно подумать про Lazy.

Еще проблема с Lazy, что ты можешь попытаться запросить этого usera, когда уже коннекшн к базе закрылся. Тогда встретимся с проблемой LazyInitException. А это плохая ошибка тк только в runtime о ней узнаем.

RinatBeybutov commented on src/main/java/com/example/libraryProject/restController/AuthController.java:
```diff
@@ -0,0 +1,47 @@
+   */
+  @PostMapping("/register")
+  public ResponseEntity<Person> register(Person person) {
+    if (userService.getUser(person).isPresent()) {
```
давай вот этот участок переделаем с использованием ExceptionHandler

RinatBeybutov commented on src/main/java/com/example/libraryProject/restController/AuthController.java:
```diff
@@ -0,0 +1,47 @@
+   */
+  @PostMapping("/login")
+  public ResponseEntity<Person> login(Person person) {
+    if (userService.getUser(person).isPresent()) {
```
и здесь тоже. Обработка ошибок сервиса не должна быть здесь, это внутренняя работа сервиса

RinatBeybutov commented on src/main/java/com/example/libraryProject/restController/BookController.java:
```diff
@@ -0,0 +1,112 @@
+   * Получение списка всех книг
+   */
+  @GetMapping
+  public List<Book> getAllBooks() {
```
куда здесь потерялось ResponseEntity? Всегда возвращаем ResponseEntity с чем=нибудь

RinatBeybutov commented on src/main/java/com/example/libraryProject/restController/BookController.java:
```diff
@@ -0,0 +1,112 @@
+   * Получение книги по её id
+   */
+  @GetMapping("/{id}")
+  public ResponseEntity<Optional<Book>> getBookById(@PathVariable Long id) {
```
Optional не нужно возвращать, ошибку нужно обработать через ExceptionHandler

https://www.baeldung.com/exception-handling-for-rest-with-spring

RinatBeybutov commented on src/main/java/com/example/libraryProject/restController/BookController.java:
```diff
@@ -0,0 +1,112 @@
+   */
+  @GetMapping("/coverBook/{id}")
+  public ResponseEntity<String> getBookCoverById(@PathVariable Long id) {
+    Book book = bookService.getBookById(id).orElse(null);
```
для чего здесь вообще нужна переменная book ?

RinatBeybutov commented on src/main/java/com/example/libraryProject/restController/BookController.java:
```diff
@@ -0,0 +1,112 @@
+  public ResponseEntity<String> assignBookToCurrentUser(@PathVariable("id") Long bookId) {
+    log.info("Received request to assign book with id {}", bookId);
+    bookService.assignBookToCurrentUser(bookId);
+    return ResponseEntity.ok("Book successfully assigned to current user");
```
текст не нужно отдавать, ResponseEntity.ok().build().
А возвращаемый тип - ResponseEntity<Void>

RinatBeybutov commented on src/main/java/com/example/libraryProject/restController/BookController.java:
```diff
@@ -0,0 +1,112 @@
+   */
+  @PreAuthorize("hasRole('ROLE_ADMIN')")
+  @PostMapping
+  public ResponseEntity<Void> createBook(@RequestBody Book book) {
```
метод создания обычно возвращает созданный объект

RinatBeybutov commented on src/main/java/com/example/libraryProject/restController/BookController.java:
```diff
@@ -0,0 +1,112 @@
+   */
+  @PreAuthorize("hasRole('ROLE_ADMIN')")
+  @PutMapping("/{id}")
+  public ResponseEntity<Void> updateBookById(@PathVariable Long id, @RequestBody Book book) {
```
метод обновления обычно возвращает обновленный объект

RinatBeybutov commented on src/main/java/com/example/libraryProject/service/BookService.java:
```diff
@@ -0,0 +1,128 @@
+    if (existingBook.isPresent()) {
+      throw new RuntimeException("Book with name '" + book.getName() + "' already exists");
+    }
+    String currentUser = UserUtils.getCurrentUsername();
```
имена у пользователей могут быть одинаковые. Намного лучше использовать id пользователя в сущности

RinatBeybutov commented on src/main/java/com/example/libraryProject/config/UserUtils.java:
```diff
@@ -0,0 +1,32 @@
+   * Получение текущего времени
+   */
+  public static LocalDateTime getCurrentTime() {
+    return LocalDateTime.now();
```
не вижу смысла выносить создание текущего времени в отдельный метод. Где нужно, там и вызывай LocalDateTime.now()

RinatBeybutov commented on src/main/java/com/example/libraryProject/service/BookService.java:
```diff
@@ -0,0 +1,128 @@
+  /**
+   * Получить книги по владельцу
+   */
+  public List<Book> getBooksByOwner(Long userid) {
```
@Transactional

RinatBeybutov commented on src/main/java/com/example/libraryProject/service/BookService.java:
```diff
@@ -0,0 +1,128 @@
+   */
+  @Transactional
+  public Optional<Book> getBookById(Long id) {
+    return Optional.ofNullable(repository.findById(id)
```
этот метод должен возвращать Book. У тебя же обрабатывается ситуация отсутствия книги

RinatBeybutov commented on src/main/java/com/example/libraryProject/service/BookService.java:
```diff
@@ -0,0 +1,128 @@
+    Person owner = personRepository.findByName(username)
+        .orElseThrow(() -> new RuntimeException("Person with name " + username + " not found"));
+
+    if (book.getOwner() != null) {
```
давай поменяем местами проверку и поиск пользователя по имени. Если книга уже у кого-то, то нет смысла искать пользователя

RinatBeybutov commented on src/main/java/com/example/libraryProject/service/BookService.java:
```diff
@@ -0,0 +1,128 @@
+   */
+  @Transactional
+  public void updateBook(Long id, Book book) {
+    if (repository.existsById(id)) {
```
давай поменяем структуру метода на такую:
if (){
throw
}

сохраняем

RinatBeybutov commented on src/main/java/com/example/libraryProject/service/BookService.java:
```diff
@@ -0,0 +1,128 @@
+   */
+  @Transactional
+  public String getBookCover(Long id) {
+    if (repository.existsById(id)) {
```
здесь нужно порефакторить и использовать метод, который получает книгу и в нем обрабатывается ошибка

RinatBeybutov commented on src/main/java/com/example/libraryProject/service/PersonService.java:
```diff
@@ -0,0 +1,45 @@
+  /**
+   * Создать пользователя
+   */
+  public void createUser(Person person) {
```
транзакция?

RinatBeybutov commented on src/main/java/com/example/libraryProject/service/PersonService.java:
```diff
@@ -0,0 +1,45 @@
+  /**
+   * Получить пользователя по имени
+   */
+  public Optional<Person> getUser(Person person) {
```
транзакция?

RinatBeybutov commented on src/main/resources/application.yml:
```diff
@@ -2,3 +2,25 @@ spring:
+
+  jpa:
+    hibernate:
+      ddl-auto: update
```
здесь none? у тебя же flyway

RinatBeybutov commented on src/main/resources/db/migration/V1.1__create_books_table.sql:
```diff
@@ -0,0 +1,14 @@
+    created_at TIMESTAMP,
+    updated_at TIMESTAMP,
+    removed_at TIMESTAMP,
+    created_user varchar(255),
```
здесь лучше id пользователя. И еще foreign key добавить

RinatBeybutov commented on src/main/resources/db/migration/V2.1__create_persons_table.sql:
```diff
@@ -0,0 +1,15 @@
+    roles VARCHAR(255),
+    created_at TIMESTAMP,
+    removed_at TIMESTAMP,
+    created_user VARCHAR(255),
```
внешний ключ на id пользователя

RinatBeybutov commented on src/test/java/com/example/libraryProject/LibraryProjectApplicationTests.java:
```diff
@@ -6,8 +6,8 @@
-	@Test
-	void contextLoads() {
-	}
+  @Test
```
удалить этот тест

RinatBeybutov commented on src/test/java/com/example/libraryProject/controller/TestBookController.java:
```diff
@@ -0,0 +1,39 @@
+  @DisplayName("Получение всех задач")
+  void testGetAllTasks() {
+
+    List<Book> books = bookController.getAllBooks();
```
это не тестирует контроллер, нужен RestTemplate

RinatBeybutov commented on src/test/java/com/example/libraryProject/repository/TestBookRepository.java:
```diff
@@ -0,0 +1,36 @@
+ *
+ * @author ITWeiss
+ */
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
```
для теста репозитория достаточно будет @DataJpaTest

ITWeiss committed:
Исправлены ошибки:
- BookService
- BookController
- AuthController
- UserUtils
- PersonService
- application.yml
- test
- db/migration

ITWeiss committed:
Исправлены ошибки:
- BookService
- BookController
- AuthController
- UserUtils
- PersonService
- application.yml
- test
- db/migration

RinatBeybutov commented on src/main/java/com/example/libraryProject/config/UserUtils.java:
```diff
@@ -23,10 +24,14 @@ public static String getCurrentUsername() {
+  public static Long getCurrentUserId() {
+    Object principal = SecurityContextHolder.getContext().getAuthentication().getPrincipal();
+    if (principal instanceof UserDetails userDetails) {
+      return ((Person) userDetails).getId();
```
а почему тогда сразу не проверить на тип Person?

RinatBeybutov commented on src/main/java/com/example/libraryProject/exception/BookNotFoundException.java:
```diff
@@ -0,0 +1,12 @@
+ */
+public class BookNotFoundException extends RuntimeException {
+  public BookNotFoundException(Long id) {
+    super("Книга с ID " + id + " не найдена");
```
StringFormat

RinatBeybutov commented on src/main/java/com/example/libraryProject/exception/UserNotFoundException.java:
```diff
@@ -0,0 +1,12 @@
+ */
+public class UserNotFoundException extends RuntimeException {
+  public UserNotFoundException(String username) {
+    super("User with name " + username + " not found");
```
StringFormat. Здесь английский, а выше - русский. Давай приведем к какому-то единому виду

RinatBeybutov commented on src/main/java/com/example/libraryProject/restController/AuthController.java:
```diff
@@ -26,7 +29,7 @@ public class AuthController {
   public ResponseEntity<Person> register(Person person) {
     if (userService.getUser(person).isPresent()) {
-      return ResponseEntity.badRequest().build();
+      throw new UserAlreadyExistsException("User with the same details already exists");
```
давай это все уберем в сервис в метод создания. А в контроллере чтобы у нас был только вызов метода сервиса

RinatBeybutov commented on src/main/java/com/example/libraryProject/restController/AuthController.java:
```diff
@@ -37,11 +40,12 @@ public ResponseEntity<Person> register(Person person) {
-      return ResponseEntity.ok(person);
-    } else {
-      return ResponseEntity.badRequest().build();
+    Optional<Person> dbPerson = userService.getUser(person);
```
тоже в сервис убираем

RinatBeybutov commented on src/main/java/com/example/libraryProject/service/BookService.java:
```diff
@@ -72,57 +78,50 @@ public void deleteBookById(Long id) {
+      throw new BookAlreadyAssignedException(book.getOwner().getName());
     }
-
+    String username = UserUtils.getCurrentUsername();
```
появился метод UserUtils.getCurrentUserId(); лучше использовать его, а потом по id найти пользователя в базе

RinatBeybutov commented on src/main/resources/db/migration/V1.1__create_books_table.sql:
```diff
@@ -7,8 +7,9 @@ CREATE TABLE books(
     updated_at TIMESTAMP,
     removed_at TIMESTAMP,
-    created_user varchar(255),
+    created_user INTEGER REFERENCES users(id),
```
created_user_id и далее также

RinatBeybutov commented on src/test/java/com/example/libraryProject/controller/TestBookController.java:
```diff
@@ -21,19 +25,23 @@
+        "http://localhost:8080/books",
+        HttpMethod.GET,
+        null,
+        new ParameterizedTypeReference<List<Book>>(){}
```
эту страшную штуку можно заменить на Book[].class

ITWeiss committed:
Исправлены ошибки:
- AuthController
- Book
- BookAlreadyAssignedException
- BookController
- BookNotFoundException
- BookService
- PersonService
- TestBookController
- UserNotFoundException
- UserUtils
- миграции

RinatBeybutov commented on src/main/java/com/example/libraryProject/entity/Book.java:
```diff
@@ -42,7 +42,7 @@ public class Book {
   private String removedPerson;
 
-  @ManyToOne(fetch = FetchType.LAZY)
+  @ManyToOne(fetch = FetchType.EAGER)
```
EAGER стоит по умолчанию, зайди в исходник этой аннотации

RinatBeybutov commented on src/main/java/com/example/libraryProject/restController/AuthController.java:
```diff
@@ -40,11 +34,8 @@ public ResponseEntity<Person> register(Person person) {
-      throw new UserNotFoundException("User not found");
-    }
-    return ResponseEntity.ok(dbPerson.get());
+    Person dbPerson = userService.getUser(person);
```
почему dbPerson ? можно просто person

RinatBeybutov commented on src/main/java/com/example/libraryProject/service/PersonService.java:
```diff
@@ -28,21 +29,19 @@ public class PersonService {
+    if (personRepository.findByName(person.getName()).isPresent()) {
+      throw new UserAlreadyExistsException("User with the same details already exists");
     }
+    person.setPassword(passwordEncoder.encode(person.getPassword()));
```
давай перепишем так:
String encodedPassword = passwordEncoder.encode(person.getPassword());
person.setPassword(encodedPassword)
..

RinatBeybutov commented on src/test/java/com/example/libraryProject/controller/TestBookController.java:
```diff
@@ -32,16 +30,14 @@ public class TestBookController extends PostgresDbTestcontainers {
   void testGetAllBooks() {
 
-    ResponseEntity<List<Book>> books = restTemplate.exchange(
+    ResponseEntity<Book[]> books = restTemplate.getForEntity(
```
вот тут можно использовать var. а можно и так оставить

RinatBeybutov commented on src/main/java/com/example/libraryProject/service/BookService.java:
```diff
@@ -0,0 +1,126 @@
+  @Transactional
+  public String getBookCover(Long id) {
+    Book book = getBookById(id);
+    return String.format("Book: %s %s. Year of publication: %d",
```
на самом деле, обложка - это картинка. В Spring есть поддержка multipart файлов. Нужно это доработать

RinatBeybutov commented on src/main/java/com/example/libraryProject/restController/BookController.java:
```diff
@@ -0,0 +1,111 @@
+   */
+  @PostMapping("/assign/{id}")
+  public ResponseEntity<Void> assignBookToCurrentUser(@PathVariable("id") Long bookId) {
+    log.info("Received request to assign book with id {}", bookId);
```
здесь есть лог, а в других методах контроллера - нет. Давай добавим, но с уровнем debug

RinatBeybutov commented on Readme.md:
```diff
@@ -1,30 +1,35 @@
+http://localhost:8080
+
+### Запуск сервера проверки JWT - токена
+http://localhost:9000
```
код этого сервера здесь есть?

RinatBeybutov commented on src/main/java/com/example/libraryProject/entity/Person.java:
```diff
@@ -0,0 +1,45 @@
+  private String role;
+  private LocalDateTime createdAt;
+  private LocalDateTime removedAt;
+  private String createdPerson;
```
здесь нужно поправить на Сущность Person и @ManyToOne
Это косяк тз, там неправильно написано

RinatBeybutov commented on src/main/java/com/example/libraryProject/entity/Book.java:
```diff
@@ -0,0 +1,48 @@
+  private LocalDateTime createdAt;
+  private LocalDateTime updatedAt;
+  private LocalDateTime removedAt;
+  private String createdPerson;
```
здесь тоже:
здесь нужно поправить на Сущность Person и @manytoone

ITWeiss committed:
Исправлены ошибки:
- AuthController
- Book
- Person
- BookController
- PersonService
- TestBookController

ITWeiss committed:
Исправлен стиль кода
